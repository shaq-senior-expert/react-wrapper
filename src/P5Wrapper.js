import React from 'react';
import p5 from 'p5';

export default class P5Wrapper extends React.Component {

  componentDidMount() {
    this.canvas = new p5(this.props.sketch, this.refs.wrapper);
    this.canvas.myCustomRedrawAccordingToNewPropsHandler(this.props);
  }

  componentWillReceiveProps(newprops) {
    if( this.canvas.myCustomRedrawAccordingToNewPropsHandler ) {
      this.canvas.myCustomRedrawAccordingToNewPropsHandler(newprops);
    }
  }

  render() {
    return <div ref="wrapper"></div>;
  }
}
