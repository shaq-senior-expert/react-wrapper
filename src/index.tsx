import p5 from "p5";
import deepEqual from "deep-equal";
import React, { createRef, FC, memo, useEffect, useState } from "react";

export interface SketchProps {
  [key: string]: any;
}

export interface Sketch {
  (instance: p5): void;
}

export interface P5WrapperProps {
  sketch: Sketch;
  [key: string]: any;
}

export interface P5Instance extends p5 {
  updateWithProps?: (props: SketchProps) => void;
}

function createCanvas(sketch: Sketch, container: HTMLDivElement) {
  return new p5(sketch, container) as P5Instance;
}

const ReactP5WrapperComponent: FC<P5WrapperProps> = ({
  sketch,
  children,
  ...props
}) => {
  const wrapper = createRef<HTMLDivElement>();
  const [instance, setInstance] = useState<P5Instance>();

  useEffect(() => {
    instance?.updateWithProps?.(props);
  }, [props]);

  useEffect(() => {
    if (wrapper.current === null) return;
    instance?.remove();
    const canvas = createCanvas(sketch, wrapper.current);
    canvas.updateWithProps?.(props);
    setInstance(canvas);
  }, [sketch, wrapper.current]);

  return <div ref={wrapper}>{children}</div>;
};

export const ReactP5Wrapper = memo(
  ReactP5WrapperComponent,
  (previousProps: P5WrapperProps, nextProps: P5WrapperProps) => {
    return deepEqual(previousProps, nextProps, { strict: true });
  }
);
