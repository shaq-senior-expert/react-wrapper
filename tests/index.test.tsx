import React from "react";
import { render } from "@testing-library/react";
import { ReactP5Wrapper } from "../src/index";
import { sketch } from "./__fixtures__/sketch";

describe("P5Wrapper", () => {
  it("renders the component", () => {
    const { container: wrapper, asFragment } = render(
      <ReactP5Wrapper sketch={sketch} />
    );
    const canvas = wrapper.querySelector("canvas");

    expect(canvas).not.toBeNull();
    expect(canvas).toBeInstanceOf(HTMLCanvasElement);
    expect(asFragment()).toMatchSnapshot();
  });
});
