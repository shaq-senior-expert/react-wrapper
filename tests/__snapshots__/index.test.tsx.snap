// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`P5Wrapper renders the component 1`] = `
<DocumentFragment>
  <div>
    <canvas
      class="p5Canvas"
      height="400"
      id="defaultCanvas0"
      style="width: 720px; height: 400px;"
      width="720"
    />
  </div>
</DocumentFragment>
`;
